METRON-1039: Add ZIP function to Stellar

cestella opened the pull request:
## Contributor Comments
Stellar could use a ZIP function i.e.
* `ZIP( [ 1, 2, 3], [ 4, 5, 6] ) == [ [1, 4], [2, 5], [3, 6] ]`
* `ZIP( [ 1, 2 ], [4, 5, 6] ) == [ [1, 4], [2, 5] ]` 
* `ZIP_LONGEST( [ 1, 2 ], [4, 5, 6] ) == [ [1, 4], [2, 5], [null, 6] ]` 

You can manually validate this in the Stellar REPL.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?


### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

ottobackwards commented on metron-stellar/stellar-common/README.md:
```diff
@@ -711,6 +713,18 @@ In the core language functions, we support basic functional programming primitiv
   * Returns: The current year
 
+### `ZIP`
+  * Description: Zips lists into a single list where the ith element is an list containing the ith items from the constituent lists.
```
Is there a link to a page explaining "ZIP" in this context, formally?

cestella replied:
Well, there is https://en.wikipedia.org/wiki/Convolution_(computer_science), what do you think?

ottobackwards replied:
That would great

simonellistonball replied:
might be kinder to link to the python docs on zip, or at least an example that isn't in lisp :)

cestella replied:
@simonellistonball Haskell example coming up! ;)

ottobackwards commented:
+1 - built + test travis style, ran stellar cli.
Nice work @cestella

nickwallen commented:
Do we have good unit tests for the function's behavior with empty lists, nulls and other boundary conditions?

cestella commented:
@nickwallen @ottobackwards A couple of modifications:
* I renamed `ZIP_JAGGED` to `ZIP_LONGEST` to conform more to python's example
* I added links to more context in the documentation
* I added some better null/empty/boundary condition checks.

simonellistonball commented on metron-stellar/stellar-common/src/test/java/org/apache/metron/stellar/dsl/functions/FunctionalFunctionsTest.java:
```diff
@@ -24,13 +24,124 @@
+            )
+    {
+      int o = (int) run(expr, variables);
+      Assert.assertEquals(1*4 + 2*5 + 3*6, o, 1e-7);
```
surely this does not actually test the structure is correct. I could get the same asserted result from zip(a, reversed(b)). I think we should have tests around defined ordering of pairs.

cestella replied:
`reduce(zip(a, reversed(b)), (s, x) -> s + x[0]*x[1], 0)` would yield `1*6 + 2*5 + 3*1 == 19` which is not `1*4 + 2*5 + 3*6 == 32`, so it is ensuring that the lists align.  The previous case involving evaluating jsut the output of `ZIP([1,2,3], [4, 5, 6])` takes care of the ordering per-element (`1` is in the first spot and `4` is in the second spot, etc)

simonellistonball replied:
good point, i should probably have looked at the numbers.

ottobackwards replied:
math fight!

ottobackwards replied:
and don't call him shirly

cestella replied:
Entirely reasonable conclusion, though @simonellistonball Good to think it through.

jjmeyer0 commented:
@cestella I was wondering what you think about having a zip function that we could pass a lambda to. It could potentially be nicer to read than combining a `ZIP` with a `REDUCE`. 

```
ZIP_WITH([1, 2, 3], [1, 2, 3], (a, b) -> a + b) == [2, 4, 6]
```

cestella commented:
@jjmeyer0 I don't have a problem with it, but can you cite any previous work (i.e. another language) that works that way?  The question I would have is, should each arg be exploded as separate args (like you have it) or should they be sent in as a single arg which is a list?

The more I think about it, the more I like the notion.  However, I recommend we make that a follow-on, because I think the semantics should probably be discussed by themselves.  Thoughts?

cestella commented:
Just a follow-up to `ZIP_WITH` with a bit more context, it appears that it's in haskell and scala was going to add it (see discussion [here](https://issues.scala-lang.org/browse/SI-1512)) but decided to hold off.

I'm propose we make a JIRA calling for `ZIP_WITH` and `ZIP_WITH_INDEX` that fits @jjmeyer0 's example code.  Further thoughts?

jjmeyer0 commented:
@cestella I will create a Jira for it and reference the above link. I agree that this ticket isn't the best place to implement this.

jjmeyer0 commented:
@cestella open ticket for ZIP_WITH and ZIP_WITH_INDEX. To make sure we are on the same page for ZIP_WITH_INDEX, can you skim over the jira when you have time?
